src/librustdoc/formats/renderer.rs

```diff
@@ -0,0 +1,116 @@
+    fn item(&mut self, item: clean::Item, cache: &Cache) -> Result<(), Error>;
+
+    /// Renders a module (should not handle recursing into children).
+    fn mod_item_in(
+        &mut self,
+        item: &clean::Item,
+        item_name: &str,
+        cache: &Cache,
+    ) -> Result<(), Error>;
+
+    /// Runs after recursively rendering all sub-items of a module.
+    fn mod_item_out(&mut self, item_name: &str) -> Result<(), Error>;
+
+    /// Post processing hook for cleanup and dumping output to files.
+    fn after_krate(&mut self, krate: &clean::Crate, cache: &Cache) -> Result<(), Error>;
+
+    /// Called after everything else to write out errors.
+    fn after_run(&mut self, diag: &rustc_errors::Handler) -> Result<(), Error>;
+}
+
+#[derive(Clone)]
+pub struct Renderer;
+
+impl Renderer {
+    pub fn new() -> Renderer {
+        Renderer
+    }
+
+    /// Main method for rendering a crate.
+    pub fn run<T: FormatRenderer + Clone>(
+        self,
+        krate: clean::Crate,
+        options: RenderOptions,
+        render_info: RenderInfo,
+        diag: &rustc_errors::Handler,
+        edition: Edition,
+    ) -> Result<(), Error> {
+        let (krate, mut cache) = Cache::from_krate(
+            render_info.clone(),
+            options.document_private,
+            &options.extern_html_root_urls,
+            &options.output,
+            krate,
+        );
+
+        let (mut format_renderer, mut krate) =
+            T::init(krate, options, render_info, edition, &mut cache)?;
+
+        let cache = Arc::new(cache);
+        // Freeze the cache now that the index has been built. Put an Arc into TLS for future
+        // parallelization opportunities
+        CACHE_KEY.with(|v| *v.borrow_mut() = cache.clone());
+
+        let mut item = match krate.module.take() {
+            Some(i) => i,
+            None => return Ok(()),
+        };
+
+        item.name = Some(krate.name.clone());
+
+        // Render the crate documentation
+        let mut work = vec![(format_renderer.clone(), item)];
+
+        while let Some((mut cx, item)) = work.pop() {
+            if item.is_mod() {
+                // modules are special because they add a namespace. We also need to
+                // recurse into the items of the module as well.
+                let name = item.name.as_ref().unwrap().to_string();
+                if name.is_empty() {
+                    panic!("Unexpected module with empty name");
+                }
+
+                cx.mod_item_in(&item, &name, &cache)?;
+                let module = match item.inner {
+                    clean::StrippedItem(box clean::ModuleItem(m)) | clean::ModuleItem(m) => m,
+                    _ => unreachable!(),
+                };
+                for it in module.items {
+                    debug!("Adding {:?} to worklist", it.name);
+                    work.push((cx.clone(), it));
```

Comment:
Cloning this for each _item_ seems like a waste :/

Reply:
Yeah I thought the same thing. It's how @GuillaumeGomez implemented it last year and I assumed it was to circumvent ownership issues, but looking at it again it should be fine to share a mutable copy (and if we ever want to parallelize item rendering we could put it behind a mutex) so I'll go ahead and remove the cloning.

Reply:
Shoot nevermind, I just remembered why this was necessary (after seeing hundreds of tests fail): the html renderer maintains a mutable "current" path which updates as it traverses submodules. It makes the output depend on the order in which items get rendered and cloning the `Context` fixes that by preserving a "snapshot" of that current path for each item. When I've got a bit of time I'll refactor the html renderer to not do that and then we can remove this cloning as well, but I'm reverting for now though.